### PR TITLE
🧹 Add operator version to status of `MondooAuditConfig`

### LINKED ISSUE #380: Add ready check for the operator verifying that all MondooAuditConfigs have been reconciled
The goal is to have the Mondoo operator stating it is ready only if all `MondooAuditConfig` CRs have been reconciled at least once by the current version of the operator.

First, we need to extend the `MondooAuditConfig.Status` and add an extra property `ReconciledByOperatorVersion` (or some other name). That property should store the version of the Mondoo operator that last reconciled the CR.

After the status is extended, the operator should implement another ready check [here](https://github.com/mondoohq/mondoo-operator/blob/main/main.go#L123). The check should list all `MondooAuditConfig` and verify that the `ReconciledByOperatorVersion` field is equal to the current version of the operator. Only if that is the case, the operator should state it is ready.

### PR DESCRIPTION
**Pull request by czunker**
This adds the version of the operator which reconciled the `MondooAuditConfig`
to the `Status` section of the CR.

The new readiness check waits for all `MondooAuditConfigs` to include this version before the operator is considered ready.

Fixes #380

Signed-off-by: Christian Zunker <christian@mondoo.com>

### CODE REVIEW

**Comment by imilchev** [comment]
> This includes changes to the `MondooAuditConfig` CRD. Do we bump the version or are there more changes that we should integrate?

This is just an update in the `Status` of the CRD which is considered an "internal" thing. Nothing really changes for the end-user. We can release this as a patch version just fine.

File: controllers/mondooauditconfig_controller.go [hunk 1]
```diff
@@ -265,6 +268,16 @@ func (r *MondooAuditConfigReconciler) Reconcile(ctx context.Context, req ctrl.Re
 		}
 	}
 
+	// Update status.ReconciledByOperatorVersion to the running operator version
+	if mondooAuditConfig.Status.ReconciledByOperatorVersion != version.Version {
+		mondooAuditConfig.Status.ReconciledByOperatorVersion = version.Version
+		err := r.Status().Update(ctx, mondooAuditConfig)
```
Comment (imilchev) [inline controllers/mondooauditconfig_controller.go:274]: You don't need to update here. The code on line 281 already performs the update if it's necessary. Check what that function does. Here you should just set the mondooAuditConfigCopy.Status.ReconciledByOperatorVersion = version.Version`

File: tests/integration/audit_config_base_suite.go [hunk 2]
```diff
@@ -300,6 +301,32 @@ func (s *AuditConfigBaseSuite) testMondooAuditConfigAdmissionMissingSA(auditConf
...
+			SkipContainerResolution: true,
+		},
+	}
+	s.Require().NoErrorf(
+		s.testCluster.K8sHelper.Clientset.Create(s.ctx, operatorConfig), "Failed to create MondooOperatorConfig")
+
+	// Enable nothing
+	zap.S().Info("Create an audit config that enables nothing.")
+	s.NoErrorf(
+		s.testCluster.K8sHelper.Clientset.Create(s.ctx, &auditConfig),
```
Comment (imilchev) [inline tests/integration/audit_config_base_suite.go:323]: it's a little strange because the function is called `testMondooAuditConfigEmpty` but the actual `MondooAuditConfig` is provided as a parameter. It actually depends on the caller of this function whether the `MondooAuditConfig` will enable anything or not

File: tests/integration/audit_config_test.go [hunk 3]
```diff
@@ -13,6 +13,11 @@ type AuditConfigSuite struct {
 	AuditConfigBaseSuite
 }
 
+func (s *AuditConfigSuite) TestReconcile_Empty() {
```
Comment (imilchev) [inline tests/integration/audit_config_test.go:16]: Is there any gain in testing this? We could just add the version check in all the other tests I suppose...

File: controllers/mondooauditconfig_controller.go (hunk 1 above)
Comment (czunker) [reply controllers/mondooauditconfig_controller.go]: Thanks. Fixed this.

File: tests/integration/audit_config_base_suite.go (hunk 2 above)
Comment (czunker) [reply tests/integration/audit_config_base_suite.go]: I changed the wording to make more clear what's happening.

I think, that the test depends on the `MondooAuditConfig` of the caller is intentional. This is how the tests were designed to make it easier to call them in different contexts.

File: tests/integration/audit_config_test.go (hunk 3 above)
Comment (czunker) [reply tests/integration/audit_config_test.go]: I now opted for both.

I'd like to keep this test to spot bugs that are independent of any resources being created.

File: pkg/health/health.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type HealthChecks struct {
+	Client client.Client
+	Log    logr.Logger
+}
+
+func (h *HealthChecks) CheckMondooAuditConfigsForOperatorVersion(req *http.Request) error {
+	ctx := context.TODO()
```
Comment (imilchev) [inline pkg/health/health.go:21]: use `context.Background()` instead of `TODO`. The `TODO` context shows up the same as TODOs left in the code.

File: pkg/health/health.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+	"go.mondoo.com/mondoo-operator/pkg/version"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type HealthChecks struct {
+	Client client.Client
+	Log    logr.Logger
+}
+
+func (h *HealthChecks) CheckMondooAuditConfigsForOperatorVersion(req *http.Request) error {
```
Comment (imilchev) [inline pkg/health/health.go:20]: I think a name like `AreAllMondooAuditConfigsReconciled` is more suitable here

File: pkg/health/health.go [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+	ctx := context.TODO()
+	mondooAuditConfigs := &v1alpha2.MondooAuditConfigList{}
+	if err := h.Client.List(ctx, mondooAuditConfigs); err != nil {
+		h.Log.Error(err, "error listing MondooAuditConfigs")
+		return err
+	}
+
+	for _, mac := range mondooAuditConfigs.Items {
+		if mac.Status.ReconciledByOperatorVersion != version.Version {
+			errorMsg := fmt.Sprintf("MondooAuditConfig not yet completly reconciled: namespace=%s MondooAuditConfig=%s", mac.Namespace, mac.Name)
```
Comment (imilchev) [inline pkg/health/health.go:30]: To make the message more descriptive you can rephrase it to something like: "MondooAuditConfig 'ns'/'name' has not been reconciled by operator 'version'." You could even state what is the current value of the `ReconciledByOperatorVersion` field

File: pkg/health/health.go [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+			h.Log.Info(errorMsg)
+			return errors.New(errorMsg)
+		} else {
+			infoMsg := fmt.Sprintf("MondooAuditconfig reconciled: namespace=%s MondooAuditConfig=%s", mac.Namespace, mac.Name)
+			h.Log.Info(infoMsg)
```
Comment (imilchev) [inline pkg/health/health.go:35]: I think this should be a debug message since it isn't really required by the end-user to know this

File: tests/integration/audit_config_base_suite.go [hunk 8]
```diff
@@ -408,3 +448,18 @@ func (s *AuditConfigBaseSuite) checkForPodInStatus(podName string) error {
 
 	return err
 }
+
+// checkForReconciledOperatorVersion Check whether the MondooAuditConfig Status contains the current operator Version after Reconcile.
+func (s *AuditConfigBaseSuite) checkForReconciledOperatorVersion() error {
```
Comment (imilchev) [inline tests/integration/audit_config_base_suite.go:453]: you could move this function somewhere in the test framework instead of defining it in the test suite. In this way it can be re-used by other suites in the future. If you choose not to do it, then you can make it `void` as it doesn't need to return the error. You can already assert in this function

File: tests/integration/audit_config_base_suite.go [hunk 9]
```diff
@@ -298,6 +308,36 @@ func (s *AuditConfigBaseSuite) testMondooAuditConfigAdmissionMissingSA(auditConf
...
+		s.testCluster.K8sHelper.Clientset.Create(s.ctx, operatorConfig), "Failed to create MondooOperatorConfig")
+
+	// Enable nothing
+	zap.S().Info("Create an audit config that enables nothing.")
+	s.NoErrorf(
+		s.testCluster.K8sHelper.Clientset.Create(s.ctx, &s.auditConfig),
+		"Failed to create Mondoo audit config.")
+
+	err := s.checkForReconciledOperatorVersion()
+	s.Assert().NoErrorf(err, "Couldn't find expected version in MondooAuditConfig.Status.ReconciledByOperatorVersion")
```
Comment (imilchev) [inline tests/integration/audit_config_base_suite.go:340]: I believe `s.Assert().NoErrorf(..)` is equivalent to `s.NoErrorf(..)` or is that not the case?

Comment (czunker) [reply tests/integration/audit_config_base_suite.go:340]: You are correct. Changed all occurrences.

File: tests/integration/audit_config_base_suite.go (hunk 8 above)
Comment (czunker) [reply tests/integration/audit_config_base_suite.go:453]: I moved all the functions similar to this one into the framework.

File: pkg/health/health.go (hunk 7 above)
Comment (czunker) [reply pkg/health/health.go:35]: As there is no debug level, I removed it completely.

File: pkg/health/health.go (hunk 6 above)
Comment (czunker) [reply pkg/health/health.go:30]: Changed.

File: pkg/health/health.go (hunk 5 above)
Comment (czunker) [reply pkg/health/health.go:20]: Changed

File: pkg/health/health.go (hunk 4 above)
Comment (czunker) [reply pkg/health/health.go:21]: Done.

**Review by imilchev — APPROVED** [review]
looks great! let's ship it

File: controllers/mondooauditconfig_controller.go [hunk 10]
```diff
@@ -108,6 +110,7 @@ func (r *MondooAuditConfigReconciler) Reconcile(ctx context.Context, req ctrl.Re
 	mondooAuditConfig := &v1alpha2.MondooAuditConfig{}
 
 	err := r.Get(ctx, req.NamespacedName, mondooAuditConfig)
+	r.MondooAuditConfig = mondooAuditConfig
```
Comment (joelddiaz) [inline controllers/mondooauditconfig_controller.go:113]: It appears to me that this is only used for test case assertions. This should be fine for a single-threaded execution model (which is how our tests currently run), but we should not rely on this for any runtime-related activities as I would expect the Reconcile() function to have to be written in a re-entrant way.
